components/core/src/clp/hash_utils.cpp

```diff
@@ -0,0 +1,107 @@
+#include "hash_utils.hpp"
+
+#include <span>
+#include <string>
+#include <vector>
+
+#include <fmt/format.h>
+#include <openssl/hmac.h>
+#include <openssl/sha.h>
+
+#include "ErrorCode.hpp"
+#include "spdlog_with_specializations.hpp"
+
+using std::span;
+using std::string;
+using std::vector;
+
+namespace clp {
+auto EvpDigestContext::digest_update(std::span<unsigned char const> input) -> ErrorCode {
+    if (m_is_digest_finalized) {
+        return ErrorCode_Unsupported;
+    }
+    if (1 != EVP_DigestUpdate(m_md_ctx, input.data(), input.size())) {
+        return ErrorCode_Failure;
+    }
+    return ErrorCode_Success;
+}
+
+auto EvpDigestContext::digest_final_ex(std::vector<unsigned char>& hash) -> ErrorCode {
+    if (m_is_digest_finalized) {
+        return ErrorCode_Unsupported;
+    }
+
+    hash.resize(SHA256_DIGEST_LENGTH);
+    unsigned int length{};
+    if (1 != EVP_DigestFinal_ex(m_md_ctx, hash.data(), &length)) {
+        return ErrorCode_Failure;
+    }
+    if (SHA256_DIGEST_LENGTH != length) {
+        return ErrorCode_Corrupt;
+    }
+    m_is_digest_finalized = true;
+    return ErrorCode_Success;
+}
+
+auto convert_hash_to_hex_string(span<unsigned char> input) -> string {
+    string hex_string;
+    for (auto const c : input) {
+        hex_string += fmt::format("{:02x}", c);
+    }
+    return hex_string;
+}
+
+auto get_hmac_sha256_hash(
+        span<unsigned char const> input,
+        span<unsigned char const> key,
+        vector<unsigned char>& hash
+) -> ErrorCode {
+    hash.resize(SHA256_DIGEST_LENGTH);
+    unsigned int hash_length{0};
+
+    if (key.size() > INT32_MAX) {
+        SPDLOG_ERROR("Input key exceeds maximum length");
```

Comment:
Ideally, we shouldn't log in a library. I think the return code already provides enough info for the caller, right?